packages/xstate-immer/src/index.ts

```diff
@@ -61,3 +63,52 @@ export function createUpdater<TContext, TEvent extends ImmerUpdateEvent>(
     type
   };
 }
+
+const IMMER_PATCH_TYPE = '@immer.patch' as const;
+
+export interface ImmerPatchEvent<K> {
+  type: typeof IMMER_PATCH_TYPE;
+  key: K;
+  patches: Patch[];
+}
+
+export const createPatch = <T, K extends keyof T>(
+  key: K
+): {
+  (currentValue: T[K], fn: (draft: Draft<T[K]>) => void): ImmerPatchEvent<K>;
```

Comment:
I believe I've asked about this in the past, but I forgot so would love a reminder - why don't we get `currentValue` automatically in a callback bur rather require it to be passed in explicitly?

Reply:
If we just did this:

```js
createPatch(userDraft => {
  userDraft.name = 'whatever';
  // ... etc.
});
```

I don't think `immer` would work, since it needs that _initial object_ to create into a proxy, and we don't provide that in `createPatch(...)` either (although we could...)

Unless I'm wrong, and maybe `immer` can create patches by performing a recipe on an "invisible" object... 🤔

Reply:
I would say that Immer is capable to mostly do its magic based on `{}` (havent tested it though). The problems with such approach could be that:
- produced patches could be different (maybe?), and not sure if they could be applied as is on the real object, although i would expect it to be possible
- if we supply a dummy object there then it wouldnt be possivle to read the real one there, so it wouldnt be possible to use its values to compute the new one. Or rather - i suppose it wouldnt be possible as i dont think Immer records read operations that could be "replayed". I might be wrong though.

Questions is if it would be possible to supply the real object lazily to this draft function so it wouldnt have to be done by the user? I suppose this whole thing (patching) is mainly relevant for distinct machines, right? Or just in more async env? It doesnt have much benefits to it if we deal with a single machine to which we have access as one could just use regular updater just fine in such scenario. Unless one would like to store patches for some kind of draft and commit later use cases? Anyway - in async scenarios this requires sync~ access to the data on which it can operate, right? So to really benefit from this in such scenarios one would have to implement data replication logic, do I understand this correctly?

Another question is - do you think its worth supporting lenses-like approach here? Like - patches produced from a different location in the context of a particular machine but later applied at different location of the destination machine?

Reply:
> so it wouldnt have to be done by the user?

I don't think this is a worthwhile goal. Think about how a user would want to update a value normally:

```js
//       vvvvv previous count!
setCount(count + 1);
```

Instead of trying to figure out the magic to make `patchUser(user => { ... })` happen, I think `patchUser(user, u => { ... })` is more explicit, unit-testable, and less magical, and we should stick with that.

Reply:
> Another question is - do you think its worth supporting lenses-like approach here? Like - patches produced from a different location in the context of a particular machine but later applied at different location of the destination machine?

What would this look like?

Reply:
Not sure - this was just a general wondering as we've discussed lense-like immer-related API in the past. 

One idea that comes to mind:
```ts
const patchUser = createPatch('user');

service.send(
  patchUser(context.user, (u) => {
    u.name = 'X';
    u.age = 42;
  })
);

{
  on: {
    [patchUser.type]: {
      actions: patchUser.assign((ctx, ev, { patch }) => {
        return {
          differentUser: patch(ctx.differentUser)
        }
      })
    }
  }
}
```

It isn't exactly that much lense-like but it decouples the source and destination locations of the patches.